File: assignment4/solution/ce2.py
import numpy as np
import numpy.random as rand
import scipy.io
import matplotlib.image as mpimg
import cv2 as cv
from triangulate import triangulate

# https://vnav.mit.edu/material/14-2viewGeometry-notes.pdf
# https://users.cecs.anu.edu.au/~hongdong/new5pt_cameraREady_ver_1.pdf 
# https://www-users.cse.umn.edu/~hspark/CSci5980/nister.pdf 

# Why do we want a five point solver when we have an eight point solver? 
# Its a minimal solver which have fewer critical surfaces

# It's much more efficient in RANSAC compared to eight point algorithm. 
# It also has better accuracy as it has checked more geometric constraints of the problem. 

# DONT WASTE YOUR TIME LIKE ME: OPENCV HAS A FUNCTION cv2.findEssentialMat() WHICH USES DAVID NISTER'S FIVE POINT ALGORITHM TO FIND ESSENTIAL MATRIX. 

def check_in_front(P1, P2s, Xs):
    largest = 0
    P2 = None
    X = None
    for i in range(0, len(P2s)):
        x1 = np.matmul(P1, Xs[i])
        x2 = np.matmul(P2s[i], Xs[i])
        inFront = 0
        for j in range(0, len(x1[2])):
            if x1[2][j] > 0 and x2[2][j] > 0:
                inFront += 1
        if inFront > largest:
            largest = inFront
            P2 = P2s[i]
            X = Xs[i]
    return P2, X

data = scipy.io.loadmat("../assignment4data/compEx2data.mat")
im1 = mpimg.imread("../assignment4data/im1.jpg")
im2 = mpimg.imread("../assignment4data/im2.jpg")
K = data['K']
[x1, x2] = data['x'][0]

hx1 = np.vstack((x1, np.ones(len(x1[0]))))
hx2 = np.vstack((x2, np.ones(len(x2[0]))))

normx1 = np.linalg.inv(K) @ hx1
normx2 = np.linalg.inv(K) @ hx2

nx1t = np.transpose(normx1)
nx2t = np.transpose(normx2)

N = 1 # The number of iterations. 
bestInliers = 0
e = 5 # The acceptable error
for i in range(0, N):
    rx1 = [nx1t[rand.randint(0, len(nx1t))][:2] for _ in range(0,5)]
    rx2 = [nx2t[rand.randint(0, len(nx2t))][:2] for _ in range(0,5)]
    E = cv.findEssentialMat(np.array(rx1), np.array(rx2), maxIters=100)[0]
    Es = [E[i:i+3] for i in range(0, len(E), 3)]
    for E in Es:
        currentInliers = 0
        ep1 = np.transpose(E @ normx1) # It looks like the lines are already normalized. 
        ep2 = np.transpose(E @ normx2)

        for i in range(0, len(ep1)):
            x1, y1, z1 = nx2t[i]
            a1, b1, c1 = ep1[i]
            dist1 = np.abs(a1 * x1 + b1 * y1 + c1) / np.sqrt(a1**2 + b1**2)
                  
            
            x2, y2, z2 = nx1t[i]
            a2, b2, c2 = ep2[i]
            dist2 = np.abs(a2 * x2 + b2 * y2 + c2) / np.sqrt(a2**2 + b2**2)
            
            if dist1 < e and dist2 < e:
                currentInliers += 1

        if currentInliers > bestInliers:
            bestInliers = currentInliers
            bestE = E

W = [[0, -1, 0], [1, 0, 0], [0, 0, 1]]
WT = np.transpose(W)
Z = [[0, 1, 0], [-1, 0, 0], [0, 0, 0]]
[U, S, V] = np.linalg.svd(bestE)
newE = U @ np.diag([1, 1, 0]) @ V
[U, S, V] = np.linalg.svd(newE)
u3 = [[row[2]]for row in U]
mu3 = [[-row[2]]for row in U]
P1 = [[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0]]
P2a = np.hstack([U @ W @ np.transpose(V), u3])
P2b = np.hstack([U @ W @ np.transpose(V), mu3])
P2c = np.hstack([U @ np.transpose(W) @ np.transpose(V), u3])
P2d = np.hstack([U @ np.transpose(W) @ np.transpose(V), mu3]) 

X1 = np.transpose(triangulate(P1, P2a, nx1t, nx2t))
X2 = np.transpose(triangulate(P1, P2b, nx1t, nx2t))
X3 = np.transpose(triangulate(P1, P2c, nx1t, nx2t))
X4 = np.transpose(triangulate(P1, P2d, nx1t, nx2t))

P2s = [P2a, P2b, P2c, P2d]
Xs = [X1, X2, X3, X4]
P2, X = check_in_front(P1, P2s, Xs)

File: assignment4/solution/ce2.m
%Solve using given fivepoint_solver from 5 random points to get the
%essential matrix. Check for inliers using ransac and this its 5 pixels
%inside. 

% Load data

data = load('../assignment4data/compEx2data.mat');
im1 = imread('../assignment4data/im1.jpg');
im2 = imread('../assignment4data/im2.jpg');

K = data.K;
x1 = data.x{1};
x2 = data.x{2};

hx1 = [x1; ones(1, size(x1, 2))];
hx2 = [x2; ones(1, size(x2, 2))];

normx1 = K\hx1;
normx2 = K\hx2;

nx1t = normx1';
nx2t = normx2';

N = 100; % Number of iterations
e = 5; % Acceptable error
n_best_inliers = 0;
best_E = [];
for i = 1:N
    indices = randperm(size(nx1t, 1), 5);
    rx1 = nx1t(indices, :);
    rx2 = nx2t(indices, :);
    E = fivepoint_solver(rx1', rx2');
    for j = 1:length(E)
        currentE = E{j};
        F = (K^-1)'*currentE*K^-1;

        ep1 = pflat(F' * hx2);
        ep1 = ep1 ./ sqrt(repmat(ep1(1, :).^2 + ep1(2, :).^2 ,[3 1]));
        ep2 = pflat(F * hx1);
        ep2 = ep2 ./ sqrt(repmat(ep2(1, :).^2 + ep2(2, :).^2 ,[3 1]));

        d1 = abs(sum(ep1.*hx1));
        d2 = abs(sum(ep2.*hx2));
        
        inliers = (d1 < 5) & (d2 < 5);
        
        nbr_of_inliers = sum(inliers(:));
        if nbr_of_inliers > n_best_inliers
            best_E = E{j};
            n_best_inliers = nbr_of_inliers;
            best_inlier = inliers;
        end 
    end
end

W = [0 -1 0; 1 0 0; 0 0 1];
Z = [0 1 0; -1 0 0; 0 0 0];
[U, ~, V] = svd(best_E);
if det(U*V') < 0
    V = -V;
end
newE = U * diag([1 1 0]) * V';
[U, S, V] = svd(newE);
u3 = U(:, 3);
mu3 = -u3;
P1 = [eye(3) zeros(3, 1)];
P2a = [U * W * V' u3];
P2b = [U * W * V' mu3];
P2c = [U * W' * V' u3];
P2d = [U * W' * V' mu3];

X1 = pflat(triangulate(P1, P2a, nx1t, nx2t))';
X2 = pflat(triangulate(P1, P2b, nx1t, nx2t))';
X3 = pflat(triangulate(P1, P2c, nx1t, nx2t))';
X4 = pflat(triangulate(P1, P2d, nx1t, nx2t))';

P2s = {P2a, P2b, P2c, P2d};
Xs = {X1, X2, X3, X4};
[P2, X] = check_in_front(P1, P2s, Xs);
hx1 = hx1(:,best_inlier==1);
hx2 = hx2(:,best_inlier==1);
X = X(:,best_inlier==1);
reprx1 = P1*X;
reprx2 = P2*X;
distances1 = pdist2(reprx1(1:2, :)', normx1(1:2,:)');
distances1 = distances1(:,1);
distances2 = pdist2(reprx2(1:2, :)', normx2(1:2,:)');
distances2 = distances2(:,1);

histogram(distances2)
%plot3(X(1,:), X(2,:), X(3,:), ".")

save('ce2data.mat','P1', 'P2', 'X', 'hx1', 'hx2', 'K');
